nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Nexus RPCs - short recap](#nexus-rpcs---short-recap)
+- [Operation identifiers](#operation-identifiers)
+  - [Caller generated identifiers](#caller-generated-identifiers)
+    - [Handler addressability](#handler-addressability)
+      - [Mapping from client to handler identifiers](#mapping-from-client-to-handler-identifiers)
+  - [Handler generated identifiers](#handler-generated-identifiers)
+    - [Cancel before started](#cancel-before-started)
+  - [Comparing the approaches](#comparing-the-approaches)
+    - [Handler Id qualification](#handler-id-qualification)
+    - [Ids as inputs](#ids-as-inputs)
+    - [Caller friendliness](#caller-friendliness)
+    - [Data sensitivity](#data-sensitivity)
+- [Proposal](#proposal)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Nexus RPCs - short recap
+
+At the fundamental level, the Nexus protocol is a synchronous RPC protocol. Asynchronous operations are modelled on top
+of a set of pre-defined RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. These asynchronous operations can be cancelled and checked for their outcomes and current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation's when it
+is ready.
+
+## Operation identifiers
+
+Since an asynchronous operation is modelled on top of synchronous RPCs, it must be addressable (identifiable) by both
+the caller and the handler. There are two approaches for generating identifiers, caller side and handler side.
+
+### Caller generated identifiers
+
+With this approach, the caller provides an Id to each of the pre-defined RPCs.
+
+The handler _should_ qualify an identifier when dealing with untrusted sources and multi-tenant applications.
+
+The following set of RPCs would be needed to model asynchronous operations using client generated Ids:
+
+- `Start(operation, id, input, callbackURL)` - handler responds inline or indicates that an async operation was started
+- `Cancel(operation, id)` - cancel operation by Id
+- `GetResult(operation, id)` - get operation result by Id
+- `Describe(operation, id)` - get operation status by Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+Example invocations:
+
+```js
+Start(
+  'charge',
+  transactionId,
+  { customerId: 'someCustomerId', amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
```

Comment:
curious what the benefit of the token is here, vs the below, which simplifies the API by not having a token:

- provide `nexus://caller.example.com/callback`
- call `DeliverResult(transactionId, result)`
- callback is made to `nexus://caller.example.com/callback` with transactionId in the query string and/or body

Reply:
Token can be used for security in case transaction ID is guessable.
Token is not part of the spec but it will likely be generated in the Temporal implementation.

Reply:
ah i see. an alternative would be signing the message